### PR TITLE
kubectl/port-forward: complete/validate/run structure

### LINKED ISSUE #16504: CLI: convert port-forward command to new CLI command structure and expose Stop and Ready Channels
To make it easier to consume from other commands, the port-forward command should implement the Complete/Validate/Run pattern implemented in newer commands. Its options structure should expose a Ready and Stop channel. 

Also messages about the forward should be output to the command's stdout instead of glog, making it possible to redirect the command output to a different output stream.

### PR DESCRIPTION
**Pull request by hodovska**
`kubectl port-forward` command is converted to a complete/validate/run kubectl command structure specified here: https://github.com/kubernetes/kubernetes/blob/master/docs/devel/kubectl-conventions.md#command-conventions
In this PR is also exposed the ready and stop channel for API consumer.

Fixes #16504

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/30222)

### CODE REVIEW

File: pkg/kubectl/cmd/portforward.go [hunk 1]
```diff
@@ -17,21 +17,33 @@ limitations under the License.
...
 	"github.com/spf13/cobra"
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/client/restclient"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
 	"k8s.io/kubernetes/pkg/client/unversioned/portforward"
 	"k8s.io/kubernetes/pkg/client/unversioned/remotecommand"
 	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
 )
 
+type PortForwardOptions struct {
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/portforward.go]: This struct should only store flag values that are set from command line. `Config`, `Client`, `PortForwarder`, `StopChannel`, `ReadyChannel` can all be instantiated as local variables inside `ForwardPorts` without changing any params.

Comment (hodovska) [inline pkg/kubectl/cmd/portforward.go]: By the specification ```<CommandName>Config``` struct (here ```PortForwardOption```) should contain all variables required for the command to run (to ease testing). Building this I was also inspired by the ```LogsOptions``` struct https://github.com/kubernetes/kubernetes/blob/master/pkg/kubectl/cmd/logs.go#L56 used as example.

**Comment by hodovska** [comment]
@k8s-bot test this issue: #28275

File: pkg/client/unversioned/portforward/portforward.go [hunk 2]
```diff
@@ -108,19 +108,20 @@ func parsePorts(ports []string) ([]ForwardedPort, error) {
...
-func New(dialer httpstream.Dialer, ports []string, stopChan <-chan struct{}, out, errOut io.Writer) (*PortForwarder, error) {
+func New(dialer httpstream.Dialer, ports []string, stopChan <-chan struct{}, readyChan chan struct{}, out, errOut io.Writer) (*PortForwarder, error) {
 	if len(ports) == 0 {
 		return nil, errors.New("You must specify at least 1 port")
 	}
 	parsedPorts, err := parsePorts(ports)
 	if err != nil {
 		return nil, err
 	}
+	readyChan = make(chan struct{})
```
Comment (ncdc) [inline pkg/client/unversioned/portforward/portforward.go]: If the caller is passing in readyChan, I don't think you'd want to throw that away and replace it with a new channel, would you?

Comment (fabianofranz) [inline pkg/client/unversioned/portforward/portforward.go]: Being overridden.

Comment (fabianofranz) [inline pkg/client/unversioned/portforward/portforward.go]: Yeah what @ncdc said.  :)

File: pkg/kubectl/cmd/portforward.go [hunk 3]
```diff
@@ -68,81 +90,102 @@ func NewCmdPortForward(f *cmdutil.Factory, cmdOut, cmdErr io.Writer) *cobra.Comm
...
+	dialer, err := remotecommand.NewExecutor(opts.Config, method, url)
 	if err != nil {
 		return err
 	}
-	fw, err := portforward.New(dialer, ports, stopChan, f.cmdOut, f.cmdErr)
+	fw, err := portforward.New(dialer, opts.Ports, opts.StopChannel, f.cmdOut, f.cmdErr)
 	if err != nil {
 		return err
 	}
+	opts.ReadyChannel = fw.Ready
```
Comment (jlowdermilk) [inline pkg/kubectl/cmd/portforward.go]: Shouldn't this assignment be reversed? You want to override the [default portforward ready channel](https://github.com/kubernetes/kubernetes/blob/master/pkg/client/unversioned/portforward/portforward.go#L123) with whatever the caller provided in PortForwardOptions, not overwrite the options.

**Comment by ncdc** [comment]
@k8s-bot e2e test this issue: #IGNORE (gce infrastructure issue)

**Comment by jlowdermilk** [comment]
LGTM

**Comment by ncdc** [comment]
cc @csrwng

File: pkg/client/unversioned/portforward/portforward.go [hunk 4]
```diff
@@ -120,7 +120,7 @@ func New(dialer httpstream.Dialer, ports []string, stopChan <-chan struct{}, out
 		dialer:   dialer,
 		ports:    parsedPorts,
 		stopChan: stopChan,
-		Ready:    make(chan struct{}),
+		Ready:    readyChan,
```
Comment (ncdc) [inline pkg/client/unversioned/portforward/portforward.go]: Could you please add a nil check before we close the channel here: https://github.com/hodovska/kubernetes/blob/9d634168d38d6c853ca3ce1510788f04a19f4a9b/pkg/client/unversioned/portforward/portforward.go#L167?

File: pkg/kubectl/cmd/portforward.go [hunk 5]
```diff
@@ -68,81 +90,102 @@ func NewCmdPortForward(f *cmdutil.Factory, cmdOut, cmdErr io.Writer) *cobra.Comm
...
+
+// RunPortForward implements all the necessary functionality for port-forward cmd.
+func (o PortForwardOptions) RunPortForward() error {
+	pod, err := o.Client.Pods(o.Namespace).Get(o.PodName)
 	if err != nil {
 		return err
 	}
 
+	if pod.Status.Phase != api.PodRunning {
+		return fmt.Errorf("Unable to execute command because pod is not running. Current status=%v", pod.Status.Phase)
```
Comment (ncdc) [inline pkg/kubectl/cmd/portforward.go]: s/Unable/unable/

File: pkg/kubectl/cmd/portforward.go [hunk 6]
```diff
@@ -68,81 +90,102 @@ func NewCmdPortForward(f *cmdutil.Factory, cmdOut, cmdErr io.Writer) *cobra.Comm
...
+
 	signals := make(chan os.Signal, 1)
 	signal.Notify(signals, os.Interrupt)
 	defer signal.Stop(signals)
 
-	stopCh := make(chan struct{}, 1)
 	go func() {
 		<-signals
-		close(stopCh)
+		close(o.StopChannel)
```
Comment (ncdc) [inline pkg/kubectl/cmd/portforward.go]: Might want to check for nil here

File: pkg/kubectl/cmd/portforward_test.go [hunk 7]
```diff
@@ -159,18 +173,38 @@ func TestPortForwardWithPFlag(t *testing.T) {
...
+				return
+			}
+			err = opts.RunPortForward()
+		}
+		cmd.Flags().Set("pod", "foo")
+
+		cmd.Run(cmd, []string{":5000", ":1000"})
+
 		if test.pfErr && err != ff.pfErr {
 			t.Errorf("%s: Unexpected exec error: %v", test.name, err)
```
Comment (ncdc) [inline pkg/kubectl/cmd/portforward_test.go]: Looks like I did some copying & pasting way back when - would you mind replacing "exec" with "portforward"?

**Comment by ncdc** [comment]
A few minor comments

**Comment by ncdc** [comment]
LGTM. @fabianofranz or @csrwng do you want to do a final pass?

**Comment by fabianofranz** [comment]
LGTM

**Comment by ncdc** [comment]
Thanks @hodovska!!!

**Comment by sttts** [comment]
:rocket:

**Pull request merged**